File: dynamice programming/87.js
class Solution {
    constructor() {
        this.mp = new Map();
    }

    solve(s1, s2) {
        if (s1 === s2) return true;
        if (s1.length !== s2.length) return false;

        const key = `${s1},${s2}`;
        if (this.mp.has(key)) {
            return this.mp.get(key);
        }

        let result = false;
        const n = s1.length;

        for (let i = 1; i < n; i++) {
            // Check swapped case
            const swapped = this.solve(s1.substring(0, i), s2.substring(n - i)) && 
                           this.solve(s1.substring(i), s2.substring(0, n - i));
            
            if (swapped) {
                result = true;
                break;
            }

            // Check non-swapped case
            const notSwapped = this.solve(s1.substring(0, i), s2.substring(0, i)) && 
                              this.solve(s1.substring(i), s2.substring(i));
            
            if (notSwapped) {
                result = true;
                break;
            }
        }

        this.mp.set(key, result);
        return result;
    }

    isScramble(s1, s2) {
        this.mp.clear();
        return this.solve(s1, s2);
    }
}

/*
How to use:
const sol = new Solution();
console.log(sol.isScramble("great", "rgeat")); // true
*/

File: dynamice programming/87.py
# Google
# MCM -> Matrix Chain Multiplication
# Scramble String
# Recur + Memo

# Tc = O(n^4)
# Sc = O(n^2)

class Solution:
    def __init__(self):
        self.mp = {}

    def solve(self, s1, s2):
        if s1 == s2:
            return True
        if len(s1) != len(s2):
            return False

        key = (s1, s2)
        if key in self.mp:
            return self.mp[key]

        result = False
        n = len(s1)
        for i in range(1, n):
            # Check for swapped condition
            swapped = self.solve(s1[:i], s2[-i:]) and self.solve(s1[i:], s2[:-i])
            if swapped:
                result = True
                break

            # Check for not swapped condition
            not_swapped = self.solve(s1[:i], s2[:i]) and self.solve(s1[i:], s2[i:])
            if not_swapped:
                result = True
                break

        self.mp[key] = result
        return result

    def isScramble(self, s1: str, s2: str) -> bool:
        self.mp.clear()
        return self.solve(s1, s2)
